scripts/benchmark_cortex.py

```diff
@@ -1,235 +1,271 @@
-    if one_at_a_time:
-        for row in rows:
-            core.addRows([row])
-    else:
-        core.addRows(rows)
-        # core.flush()
-
-def _getTufosByIdens(idens, core):
-    core.getTufosByIdens(idens)
-
-def _getTufoByPropVal(propvals, core):
-    for p, v in propvals:
-        core.getTufoByProp(p, v)
-
-def random_normal(avg):
-    ''' Returns a number with normal distribution around avg, the very fast way '''
-    return random.randint(1, avg) + random.randint(0, avg + 1)
-
-def random_string(avg):
-    num_letters = random_normal(avg)
-    return ''.join(chr(random.randint(ord('a'), ord('a') + 25)) for x in range(num_letters))
-
-small_count = 0
-medium_count = 0
-large_count = 0
-huge_count = 0
-
-def random_val_len():
-    global small_count, medium_count, large_count, huge_count
-    x = random.random()
-    prob = SMALL_VAL_RATE
-    if x < prob:
-        small_count += 1
-        return SMALL_VAL_BYTES
-    prob += MEDIUM_VAL_RATE
-    if x < prob:
-        medium_count += 1
-        return MEDIUM_VAL_BYTES
-    prob += LARGE_VAL_RATE
-    if x < prob:
-        large_count += 1
-        return LARGE_VAL_BYTES
-    huge_count += 1
-    return HUGE_VAL_BYTES
-
-def gen_random_form():
-    num_props = random_normal(AVG_PROPS_PER_TUFO)
-    props = [random_string(AVG_PROP_NAME_LEN) for x in range(num_props)]
-    return props
-
-def gen_random_tufo(form):
-    iden = hexlify(random.bytes(16)).decode('utf8')
-    props = {}
-    for propname in form:
-        if random.random() <= INTEGER_VAL_RATE:
-            val = random.randint(-2 ** 62, 2 ** 63)
-        else:
-            val = random_string(random_val_len())
-        props[propname] = val
-    return (iden, props)
-
-def _rows_from_tufo(tufo):
-    timestamp = random.randint(1, 2 ** 63)
-    rows = []
-    iden = tufo[0]
-    for p, v in tufo[1].items():
-        rows.append((iden, p, v, timestamp))
-    return rows
-
-def flatten(iterable):
-    return list(itertools.chain.from_iterable(iterable))
-
-def _prepopulate_core(core, rows):
-    core.addRows(rows)
-
-def nth(iterable, n):
-    "Returns the nth item or a default value"
-    return next(itertools.islice(iterable, n, None))
-
-def get_random_keyval(d):
-    i = random.randint(0, len(d))
-    key = nth(d.keys(), i)
-    return (key, d[key])
+import gc
+import time
+import random
+import asyncio
+import contextlib
+import collections
+import logging
+import statistics
+
+'''
+Benchmark cortex operations
+
+TODO:  separate client process, multiple clients
+TODO:  tagprops, regex, control flow, node data, multiple layers, spawn option, remote layer
+'''
+
+logger = logging.getLogger(__name__)
+
+from typing import List, Dict, AsyncIterator, Tuple, Any, Callable, TYPE_CHECKING
+if TYPE_CHECKING:
+    import synapse.telepath as s_telepath
+    import synapse.cortex as s_cortex
+
+import synapse.common as s_common
+
+import synapse.tests.utils as s_t_utils
+
+s_common.setlogging(logger, 'ERROR')
```

Comment:
We should probably check if `__debug__` is set and log an error that the user is running un-optimized codez.